src/lib/zexe_backend/zexe_backend_common/plonk_dlog_keypair.ml

```diff
@@ -0,0 +1,198 @@
+module type Inputs_intf = sig
+  open Intf
+
+  val name : string
+
+  module Rounds : Pickles_types.Nat.Intf
+
+  module Constraint_system : T0
+
+  module Urs : sig
+    type t
+
+    val read : string -> t
+
+    val write : t -> string -> unit
+
+    val create :
+      Unsigned.Size_t.t -> Unsigned.Size_t.t -> Unsigned.Size_t.t -> t
+  end
+
+  module Index : sig
+    type t
+
+    val delete : t -> unit
+
+    val create : Constraint_system.t -> Unsigned.Size_t.t -> Urs.t -> t
+  end
+
+  module Curve : sig
+    module Affine : sig
+      type t
+    end
+  end
+
+  module Poly_comm : sig
+    module Backend : T0
+
+    type t = Curve.Affine.t Poly_comm.t
+
+    val of_backend : Backend.t -> t
+  end
+
+  module Scalar_field : T0
+
+  module Verifier_index : sig
+    type t
+
+    val create : Index.t -> t
+
+    val sigma_comm_0 : t -> Poly_comm.Backend.t
+
+    val sigma_comm_1 : t -> Poly_comm.Backend.t
+
+    val sigma_comm_2 : t -> Poly_comm.Backend.t
+
+    val ql_comm : t -> Poly_comm.Backend.t
+
+    val qr_comm : t -> Poly_comm.Backend.t
+
+    val qo_comm : t -> Poly_comm.Backend.t
+
+    val qm_comm : t -> Poly_comm.Backend.t
+
+    val qc_comm : t -> Poly_comm.Backend.t
+
+    val rcm_comm_0 : t -> Poly_comm.Backend.t
+
+    val rcm_comm_1 : t -> Poly_comm.Backend.t
+
+    val rcm_comm_2 : t -> Poly_comm.Backend.t
+
+    val psm_comm : t -> Poly_comm.Backend.t
+
+    val add_comm : t -> Poly_comm.Backend.t
+
+    val mul1_comm : t -> Poly_comm.Backend.t
+
+    val mul2_comm : t -> Poly_comm.Backend.t
+
+    val emul1_comm : t -> Poly_comm.Backend.t
+
+    val emul2_comm : t -> Poly_comm.Backend.t
+
+    val emul3_comm : t -> Poly_comm.Backend.t
+
+    val r : t -> Scalar_field.t
+
+    val o : t -> Scalar_field.t
+  end
+end
+
+module Make (Inputs : Inputs_intf) = struct
+  open Core_kernel
+  open Inputs
+
+  type t = Index.t
+
+  let name = sprintf "%s_%d_v2" name (Pickles_types.Nat.to_int Rounds.n)
+
+  let set_urs_info, load_urs =
+    let urs_info = Set_once.create () in
+    let urs = ref None in
+    let degree = 1 lsl Pickles_types.Nat.to_int Rounds.n in
+    (* TODO *)
+    let public_inputs = Unsigned.Size_t.of_int 0 in
```

Comment:
Can you have public_inputs and size be failwiths until we figure out good values?